java-client/src/main/java/com/wavefront/integrations/WavefrontDirectSender.java

```diff
@@ -0,0 +1,270 @@
+package com.wavefront.integrations;
+
+import com.wavefront.api.DataIngesterAPI;
+import com.wavefront.common.NamedThreadFactory;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.http.HttpClientConnection;
+import org.apache.http.HttpException;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpResponse;
+
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.protocol.HttpContext;
+import org.apache.http.protocol.HttpRequestExecutor;
+import org.jboss.resteasy.client.jaxrs.ClientHttpEngine;
+import org.jboss.resteasy.client.jaxrs.ResteasyClient;
+import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
+import org.jboss.resteasy.client.jaxrs.ResteasyWebTarget;
+import org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient4Engine;
+import org.jboss.resteasy.plugins.interceptors.encoding.AcceptEncodingGZIPFilter;
+import org.jboss.resteasy.plugins.interceptors.encoding.GZIPDecodingInterceptor;
+import org.jboss.resteasy.spi.ResteasyProviderFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import javax.ws.rs.core.Response;
+
+/**
+ * Wavefront Client that sends data directly to Wavefront via the direct ingestion APIs.
+ *
+ * @author Vikram Raman (vikram@wavefront.com)
+ */
+public class WavefrontDirectSender implements WavefrontSender, Runnable {
+
+  private static final String DEFAULT_SOURCE = "wavefrontDirectSender";
+  private static final Logger LOGGER = LoggerFactory.getLogger(WavefrontDirectSender.class);
+  private static final String quote = "\"";
+  private static final String escapedQuote = "\\\"";
+
+  private final Object pointsMutex = new Object();
+  private final ScheduledExecutorService scheduler;
+  private final String server;
+  private final String token;
+  private DataIngesterAPI directService;
+  private CloseableHttpClient httpClient;
+  private Future<?> scheduledFuture;
+
+  //TODO: bounded buffer?
+  private List<String> buffer = new ArrayList<>();
```

Comment:
There should probably be a limit since a machine with connectivity issue or an invalid token should not collect points forever.

Reply:
This needs to be marked volatile